### PR TITLE
Add a "--conservative" flag for "rush generate"

### PR DESCRIPTION
**Pull request by pgonzal**
Scenario:

1. A person adds a new dependency to their project's `package.json`
2. This requires them to run `rush generate`

Normally this also forces an upgrade of all SemVer range dependencies in shrinkwrap.yaml, which frequently requires the person to fix compatibility issues unrelated to their change.

This is unnecessary for repos that already handle these upgrades automatically in a separate branch.  This PR introduces a new switch "--conservative" that causes PNPM to upgrade only the minimal set of dependencies required to satisfy your `package.json` changes.

**NOTE:** This change requires pnpm-1.35.2 with the fix for https://github.com/pnpm/pnpm/issues/1054

### CODE REVIEW

**Comment by pgonzal** [comment]
@sboylems @alinsirbu FYI

File: apps/rush-lib/src/cli/actions/GenerateAction.ts [hunk 1]
```diff
@@ -110,21 +119,24 @@ export default class GenerateAction extends BaseRushAction {
...
-        console.log(os.EOL + 'Deleting ' + committedShrinkwrapFilename);
-        fsx.unlinkSync(committedShrinkwrapFilename);
-      }
-      if (fsx.existsSync(tempShrinkwrapFilename)) {
-        fsx.unlinkSync(tempShrinkwrapFilename);
+      if (this._conservativeParameter.value) {
+        if (fsx.existsSync(committedShrinkwrapFilename)) {
+          console.log(os.EOL + 'Preserving ' + committedShrinkwrapFilename);
+        } else {
+          console.log(os.EOL + 'Missing ' + committedShrinkwrapFilename);
```
Comment (nickpape-msft) [inline apps/rush-lib/src/cli/actions/GenerateAction.ts]: >nsole.log(os.EOL + 'Missing ' + committedShrinkwrapFilename [](start = 12, length = 59)

If you are missing the committedShrinkwrapFile, does doing a `--conservative` generate even make sense?

File: apps/rush-lib/src/cli/actions/GenerateAction.ts [hunk 2]
```diff
@@ -110,21 +119,24 @@ export default class GenerateAction extends BaseRushAction {
...
+        }
       }
```
[the remark this answers is not in the document]
Comment (nickpape-msft) [reply apps/rush-lib/src/cli/actions/GenerateAction.ts]: this should be an error case

---
In reply to: [171737984](https://github.com/Microsoft/web-build-tools/pull/549#discussion_r171737984) [](ancestors = 171737984)

File: apps/rush-lib/src/cli/actions/GenerateAction.ts [hunk 3]
```diff
@@ -110,19 +120,29 @@ export default class GenerateAction extends BaseRushAction {
...
+            + committedShrinkwrapFilename);
+        } else {
+          throw new Error('The "--conservative" flag cannot be used because the shrinkwrap file is missing: '
+            + committedShrinkwrapFilename);
+        }
+
+        // Copy common\config\rush\shrinkwrap.yaml --> common\temp\shrinkwrap.yaml
+        console.log(os.EOL + 'Updating ' + tempShrinkwrapFilename);
+        fsx.copySync(committedShrinkwrapFilename, tempShrinkwrapFilename);
+      } else {
```
Comment (nickpape-msft) [inline apps/rush-lib/src/cli/actions/GenerateAction.ts]: >else { [](start = 8, length = 6)

not unlinking committed shrinkwrap file?

Comment (nickpape-msft) [reply apps/rush-lib/src/cli/actions/GenerateAction.ts]: we discussed this offline and it is ok

---
In reply to: [171738895](https://github.com/Microsoft/web-build-tools/pull/549#discussion_r171738895) [](ancestors = 171738895)

File: apps/rush-lib/src/cli/actions/GenerateAction.ts (hunk 1 above)
Comment (pgonzal) [reply apps/rush-lib/src/cli/actions/GenerateAction.ts]: No, fixed

---
In reply to: [171737620](https://github.com/Microsoft/web-build-tools/pull/549#discussion_r171737620) [](ancestors = 171737620)

File: apps/rush-lib/src/cli/actions/GenerateAction.ts (hunk 3 above)
Comment (pgonzal) [reply apps/rush-lib/src/cli/actions/GenerateAction.ts]: I documented it in a separate change log.

---
In reply to: [171738956](https://github.com/Microsoft/web-build-tools/pull/549#discussion_r171738956) [](ancestors = 171738956,171738895)

**Pull request merged**